step2 - 문자열 덧셈 계산기

be-93 opened the pull request:
## 안녕하세요. 1주차 로또-TDD 미션 참여자 최웅석입니다. 😊

> 2단계 학습 문자열 덧셈 계산기 미션 리뷰 요청 드립니다.
> 제가 git 사용에 아직 미숙해서 이것저것 만져버렸더니 어제께 커밋 내용이 다시 살아나버렸습니다...  😱 
> 혹시 문제가 된다면 고칠수 있는 방법이 있을지... 알려주시면 감사하겠습니다. 😂 

### ❓ ❓ 2단계 학습하며 궁금한 점

> 

1. 저는 상수를 클래스로 관리하게 구현하였는데 리뷰어 님께서는
  상수 관련해서 관리를 하실때 보통 어떻게 하시는지 궁금합니다.

2. 이번에 TDD에 더 집중 하다보니 나중에 테스트 코드가 너무 많아지는 경험을 했습니다.😭😭
최종 결과물을 만들어내도록 리팩토링이 끝나면 기존 테스트 코드는 지우는게 좋을까요??
뭔가 코드 리뷰하시는 분들이 보기에 코드가 정신없을거 같아서 질문드립니다😵😵

저번 리뷰도 큰 도움이 됐습니다. 테스트 케이스에 의미없는 throw 를 작성하지 말라고 하셔서 
이번에는 조심해서 코딩한 것 같습니다. 🤘
감사합니다. :)

be-93 committed:
test: 요구사항1 테스트 케이스 생성

"1,2" split -> {"1","2"}
"1" split -> {"1"}

be-93 committed:
test: 요구사항2 테스트 케이스 생성

"(1,2)" substring -> "1,2"

be-93 committed:
test: 요구사항3 테스트 케이스 생성

be-93 committed:
test: Set Collection 요구사항1 테스트 케이스 생성

be-93 committed:
test: Set Collection 요구사항2 테스트 케이스 생성

be-93 committed:
test: Set Collection 요구사항3 테스트 케이스 생성

be-93 committed:
test: 공식 문서 조금 따라해보기

be-93 committed:
Merge remote-tracking branch 'origin/ungseokchoi' into ungseokchoi

be-93 committed:
test: StringAddCalculatorTest 생성

1.기본 regex 를 이용한 문자열 분리 후 숫자 합산
2.커스텀 regex 를 이용한 문자열 분리 후 숫자 합산

be-93 committed:
test: StringAddCalculatorTest 테스트 케이스 추가

1. 숫자 형식 검증

be-93 committed:
feat: CalculatorConstants 상수 클래스 생성

be-93 committed:
refactor: StringAddCalculatorTest

직접 사용할 메소드 구조 형식으로 변경

be-93 committed:
feat: StringAddCalculator 최종 생성

leeyohan93 commented on README.md:
```diff
@@ -6,4 +6,10 @@
+### 기능 요구사항
+- 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환 (예: “” => 0, "1,2" => 3, "1,2,3" => 6, “1,2:3” => 6)
+- 앞의 기본 구분자(쉼표, 콜론)외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열 앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다. 예를 들어 “//;\n1;2;3”과 같이 값을 입력할 경우 커스텀 구분자는 세미콜론(;)이며, 결과 값은 6이 반환되어야 한다.
+- 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw한다.
```
요구사항 문서 잘 작성해주셨습니다 👍
소소한 코멘트을 남겨보자면 주어진 요구사항을 기반으로 to-do list를 만들어보는 것 또한 과정을 진행하면서 도움이 많이 되실 것이라 생각합니다 :)

leeyohan93 commented on src/main/java/lotto/constants/CalculatorConstants.java:
```diff
@@ -0,0 +1,7 @@
+package lotto.constants;
+
+public class CalculatorConstants {
+
+    public final static String DEFAULT_SPLIT_REGEX = ",|:";
```
>저는 상수를 클래스로 관리하게 구현하였는데 리뷰어 님께서는
상수 관련해서 관리를 하실때 보통 어떻게 하시는지 궁금합니다.

저는 상수를 관리할 때, 따로 상수 전용 클래스를 선언하지 않고 상수가 사용되는 가장 가까운 곳에 배치하여 사용하고 있습니다.
보통 상수가 사용되는 가장 가까운 곳이라 하면 상수와 비즈니스적으로 밀접한 관련이 있는 클래스이며,
만약 public으로 여러 곳에서 참조해야할 경우는 가장 관련 있는 클래스 내부에 두고 관리하고 있습니다 :)

상수를 모아두는 클래스를 만드는 것 보다는 좀 더 관련있는 클래스 내부에 배치해주시면 좋을 것 같아요 :)

be-93 replied:
평소에 정말 고민했던 내용인데 너무 이해가 잘되는 피드백인거 같습니다. 감사합니다. 👍

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+
+public class StringAddCalculator {
+
+    private static Pattern pattern = Pattern.compile("//(.)\n(.*)");
```
생성 비용이 큰 Pattern 객체를 사용한 캐싱 💯
다만 final이 없다면 어떤 상황이 일어날 수 있을까요?

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void stringCalculator(String inputString) {
+        int sum = getNumberList(inputString);
+        System.out.println("합산 : " + sum);
```
요구사항에 필요하지 않은 로깅은 제거해주시는 것이 좋을 것 같아요  :)

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+    public int getNumberList(String inputString) {
+        String args = getCheckInputString(inputString);
+        List<Integer> list = splitInputString(args);
+        return list.stream().mapToInt(Integer::intValue).sum();
```
```suggestion
        return list.stream()
                .mapToInt(Integer::intValue)
                .sum();
```
하나의 라인에 하나의 .을 사용하면 가독성이 좋아집니다 :)
반대로 하나의 라인에 .을 사용하지 않은 경우는 사용하도록 변경해주시면 좋을 것 같아요 👍

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+        String joinString = String.join(",", strings);
+        if (!isNumber(joinString)) {
+            System.out.println("[ERROR] 잘못된 입력 값입니다.");
+            throw new RuntimeException("[ERROR] 잘못된 입력 값입니다.");
```
RuntimeException은 너무 추상적인 예외인 것 같습니다.
해당 부분에 더욱 어울리는 예외는 없을까요 ? 🤔

leeyohan93 commented on src/test/java/lotto/utils/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,145 @@
+    @ValueSource(strings = "//,\n1,2,3")
+    @DisplayName("문자열을 커스텀 regex 를 이용하여 분리 후 숫자를 합산")
+    public void getCustomRegexSumNumber(String inputString) {
+        Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(inputString);
+        String customRegex = ",|:";
+        if (matcher.find()) {
+            customRegex = matcher.group(1);
+            inputString = matcher.group(2);
+        }
+        String[] strings = inputString.split(customRegex);
+        int sum = Arrays.asList(strings).stream().mapToInt(Integer::parseInt).sum();
+
+        assertThat(sum).isEqualTo(6);
```
이전에 드렸던 피드백 확인 부탁드릴게요 :)
https://github.com/next-step/java-lotto-pro/pull/19#discussion_r740264815

be-93 replied:
given when then 의 내용이 조금 헷갈리는데 😱
given (setup) 
when (setup 된 데이터를 활용)
then (데이터가 정상적인지 확인)

이렇게 이해하면 될까요???

leeyohan93 replied:
네 맞습니다. 조금 더 부연설명을 붙여보자면 
when 에서는 setup된 데이터를 활용하여 테스트하고자 하는 메서드, 또는 행위를 실행하고
then 에서는 when에 대한 결과와 기댓값을 비교하고 검증하는 역할로 생각해주시면 좋을 것 같습니다 :)

leeyohan93 commented on src/test/java/lotto/utils/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,145 @@
+    @ParameterizedTest
+    @ValueSource(strings = {"//;\n1;2;3;4", "//;\n-1;2;3;4", "//;\n가;2;3;4" ,""})
+    public void stringCalculator(String inputString) {
+        int sum = getNumberList(inputString);
+        System.out.println("합산 : " + sum);
```
검증은 항상 assertion을 사용해주세요 :)

leeyohan93 commented on src/test/java/lotto/utils/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,145 @@
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+public class StringAddCalculatorTest {
```
>이번에 TDD에 더 집중 하다보니 나중에 테스트 코드가 너무 많아지는 경험을 했습니다.😭😭
최종 결과물을 만들어내도록 리팩토링이 끝나면 기존 테스트 코드는 지우는게 좋을까요??
뭔가 코드 리뷰하시는 분들이 보기에 코드가 정신없을거 같아서 질문드립니다😵😵

TDD 싸이클을 경험하시면서 만들어가는 모습이 매우 좋습니다 👍 
최종 결과물에 대한 구현 이후 테스트 코드들을 남겨두어야 하는 지에 대해 질문 주셨는데요.
테스트 코드는 설계 뿐만 아니라 메서드의 변경이나 리팩터링에도 큰 역할을 가지기 때문에 지우지 않는 것이 좋습니다.
다만 이러한 고민에 대한 원인은 현재 테스트하고 있는 대상이 StringAddCalculator이 아니라 StringAddCalculator의 내부 구현 로직이기 때문에 생기는 복잡도 때문인 것 같아요.
요구사항 힌트에 정의된 `TestCase 소스 코드` 를 참고하시어 내부 구현로직이 아닌 StringAddCalculator를 테스트하도록 진행해보시면 좋을 것 같습니다.

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+        return list.stream().mapToInt(Integer::intValue).sum();
+    }
+
+    public String getCheckInputString(String inputString) {
```
객체의 메서드는 꼭 필요한 메서드의 접근제어자만 public으로 설정해주는 것이 좋습니다.
객체 내부에서 사용되는 메서드라면 private으로 설정해주세요 :)

leeyohan93 reviewed (changes requested):
안녕하세요 웅석님 😄 
문자열 덧셈 계산기 구현을 잘 해주셨습니다 👍 
질문 주신 부분에 대한 답변과 소소한 코멘트들을 남겨 놓았습니다.
다음 단계로 넘어가기 전에 요구사항에 졍의된 TestCase 소스 코드를 사용해보는 것도 좋은 경험일 것 같습니다.

커밋에 대해서는 관계 없을 것 같으니 걱정 안하셔도 될 것 같습니다 :)

be-93 committed:
refactor: 피드백 수정 반영

1. 한 라인에 점은 하나만 사용한다.
2. Exception 은 추상적이지 않게 작성하도록 한다.

be-93 committed:
feat: 구현기능 목록 추가 작성

be-93 committed:
refactor: 피드백 수정 반영

1. 상수는 final 을 사용하여 불변을 보장받아야한다.
2. 필요한 메소드 외에 접근제어자를 활용해 안전하게 보호한다.

be-93 committed:
fix: validateNumberFormat 오류 수정

1.테스트 하며 수정했던 부분 미반영되어 수정

be-93 committed:
refactor: Test 클래스 피드백 수정 반영

1. 최종 클래스를 이용하여 테스트를 진행한다.
2. given when then 을 준수하다.

be-93 committed:
remove: 상수 위치 변경으로 인한 상수클래스 제거

1. 상수는 다양한 곳에서 활용하지 않고
특정한 클래스에서만 사용될 경우 클래스의 분리보다
해당 클래스에 작성하는것이 좋다.
DEFAULT_SPLIT_REGEX
CalculatorConstants -> StringAddCalculator

be-93 committed:
refactor: 상수 위치 변경으로 인한 import 제거

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+    public int getNumberList(String inputString) {
+        String args = getCheckInputString(inputString);
+        List<Integer> list = splitInputString(args);
+        return list
+                .stream()
+                .mapToInt(Integer::intValue)
+                .sum();
```
```suggestion
        return list.stream()
                .mapToInt(Integer::intValue)
                .sum();
```
한 라인에 점을 하나 쓰도록 한다면 위와 같은 형태가 될 수 있을 것 같아요! :)

leeyohan93 commented on src/test/java/lotto/utils/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,93 @@
+    @ParameterizedTest
+    @ValueSource(strings = "//,\n1,2,3")
+    @DisplayName("문자열을 커스텀 regex 를 이용하여 분리 후 숫자를 합산")
+    public void getCustomRegexSumNumber(String inputString) {
+        // given
+        String customRegex = ",|:";
+        Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(inputString);
+        if (matcher.find()) {
+            customRegex = matcher.group(1);
+            inputString = matcher.group(2);
+        }
+        String[] strings = inputString.split(customRegex);
+
+        // when
+        int sum = Arrays.asList(strings).stream().mapToInt(Integer::parseInt).sum();
+
+
+        // then
+        assertThat(sum).isEqualTo(6);
+    }
```
아직 테스트에서 StringAddCalculator가 아닌 내부 구현 코드를 테스트하고 있는 부분이 있네요..!

leeyohan93 commented on src/main/java/lotto/utils/StringAddCalculator.java:
```diff
@@ -0,0 +1,76 @@
+        int sum = getNumberList(inputString);
+    }
+
+    public int getNumberList(String inputString) {
```
해당 부분 잘 구현해주셨습니다. 소소한 코멘트를 드리자면

> 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환
> 

`StringAddCalculator` 는 위의 요구사항을 만족하기 위해 아래 세가지 책임을 가지고 있습니다.

1. 구분자를 기준으로 문자열을 구분한다.
2. 문자열을 솟쟈로 변환한다.
3. 숫자의 합을 구한다.

구현해주신 코드에서 메서드 분리를 통해 충분히 가독성있고 좋아 보이지만,생각보다 많은 책임을 수행하고 있을 수 있으며 이에 대한 각각의 단위테스트는 수행하기 힘든 형태입니다.따라서 다음 로또 미션에서는 책임의 분배와 클래스를 작은 단위로 분리 하는 것에 대해 생각하고 구현해보신다면 앞으로의 미션에도 많은 도움이 될 것이라 생각이 됩니다 :)


추가적으로 기존의 메서드 명 에서는 숫자를 sum한다는 생각을 하지 못하고, 문자열을 List<Integer>로 변환하여 반환할 것이이다 라고 생각이 드는데요!  🤔  좀 더 적합한 메서드 명이 필요한 때인 것 같아요 :)
```suggestion
    public int splitAndSum(String inputString) {
```

be-93 replied:
단위 테스트가 편한 상태로 코딩하는게 쉽지 않아서 고민이네요😭😭😭
근데 리뷰어님이 남겨주신 코멘트를 보고 머리가 조금 띵 했습니다!!!
클래스와 메소드의 책임을 바라보는 시선으로 코딩한다면 뭔가 개선될것 같습니다.🤔

leeyohan93 reviewed (approved):
안녕하세요 웅석님 😄
피드백을 잘 반영해주셨습니다!
소소한 코멘트를 남겨드렸으니 다음 단계 진행시 참고 부탁드릴게요!
이번 단계는 머지하겠습니다. 남은 단계도 화이팅입니다 :)
